Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,2354 @@
+        X_S.p(),
+        getAxis(X_S, direction));
+}
+
+// Helper function for flipping the sign such that positive value is outside of
+// the surface.
+Real calcSurfaceValue(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceValue(point_S);
+}
+// Flipping sign...
+Vec3 calcSurfaceGradient(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceGradient(point_S);
+}
+// Flipping sign...
+Mat33 calcSurfaceHessian(const ContactGeometry& geometry, const Vec3& point_S)
+{
+    return -geometry.calcSurfaceHessian(point_S);
+}
+
+} // namespace
+
+//==============================================================================
+//                          CABLE SUBSYSTEM
+//==============================================================================
+
+bool CableSubsystem::isInstanceOf(const Subsystem& s)
+{
+    return Impl::isA(s.getSubsystemGuts());
+}
+
+const CableSubsystem& CableSubsystem::downcast(const Subsystem& s)
+{
+    assert(isInstanceOf(s));
+    return static_cast<const CableSubsystem&>(s);
+}
+
+CableSubsystem& CableSubsystem::updDowncast(Subsystem& s)
+{
+    assert(isInstanceOf(s));
+    return static_cast<CableSubsystem&>(s);
+}
+
+CableSubsystem::CableSubsystem()
+{
+    adoptSubsystemGuts(new Impl());
+}
+
+CableSubsystem::CableSubsystem(MultibodySystem& mbs)
+{
+    adoptSubsystemGuts(new Impl());
+    mbs.adoptSubsystem(*this);
+}
+
+int CableSubsystem::getNumCables() const
+{
+    return getImpl().getNumCables();
+}
+
+const CableSpan& CableSubsystem::getCable(CableSpanIndex ix) const
+{
+    return getImpl().getCable(ix);
+}
+
+CableSpan& CableSubsystem::updCable(CableSpanIndex ix)
+{
+    return updImpl().updCable(ix);
+}
+
+//------------------------------------------------------------------------------
+//                        CABLE SUBSYSTEM IMPL
+//------------------------------------------------------------------------------
+
+const CableSubsystem::Impl& CableSubsystem::getImpl() const
+{
+    return SimTK_DYNAMIC_CAST_DEBUG<const Impl&>(getSubsystemGuts());
+}
+
+CableSubsystem::Impl& CableSubsystem::updImpl()
+{
+    return SimTK_DYNAMIC_CAST_DEBUG<Impl&>(updSubsystemGuts());
+}
+
+int CableSubsystem::Impl::calcDecorativeGeometryAndAppendImpl(
+    const State& state,
+    Stage stage,
+    Array_<DecorativeGeometry>& decorations) const
+{
+    if (stage != Stage::Position)
+        return 0;
+
+    for (const CableSpan& cable : cables) {
+        int returnValue = cable.getImpl().calcDecorativeGeometryAndAppend(
+            state,
+            stage,
+            decorations);
+        if (returnValue != 0) {
+            return returnValue;
+        }
+    }
+    return 0;
+}
+
+int CableSubsystem::Impl::realizeSubsystemTopologyImpl(State& state) const
+{
+    // Briefly allow writing into the Topology cache; after this the
+    // Topology cache is const.
+    Impl* wThis = const_cast<Impl*>(this);
```

Comment:
And there's no way of doing this without `const_cast`ing the system? Seems like a design flaw that a `const` realization method on a subsystem is going to mutate the subsystem.

Reply:
The virtual `Subsystem::Guts::realizeSubsystemTopologyImpl(State&) const` is const, and I need to store the cache entry indexes in the model. Not sure how to do that without mutating the model, perhaps Atomics. I see this pattern in several places, here for example https://github.com/simbody/simbody/blob/d98f05fa1899633b544757e6c179562940f0849d/Simbody/src/ExponentialSpringForce.cpp#L276
So I guess no concurrent realizing of topologies? :)